core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -264,7 +264,11 @@ class LogSegment(val log: FileRecords,
     var lastIndexEntry = 0
     maxTimestampSoFar = RecordBatch.NO_TIMESTAMP
     try {
-      for (batch <- log.batches(maxMessageSize).asScala) {
+      // record batches can grow during compaction (see KAFKA-5316), so we allow some extra space when reading
+      // the batches. Checking the message size is only used as a way to check corruption before checking the CRC
+      val batches = log.batches(maxMessageSize * 2).asScala
```

Comment:
I thought this was a safety measure to prevent an OOM if the end of the log somehow got corrupted. I'm not sure what scenarios were in mind when the check was added, but it seems reasonable to preserve it if we can find a suitable upper bound. I'm guessing your discomfort is the fact that the multiplier is chosen somewhat arbitrarily?